cpp/benchmarks/scan.cc

```diff
@@ -0,0 +1,56 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include <arrow/dataset/api.h>
+#include <fmt/format.h>
+
+#include <argparse/argparse.hpp>
+#include <cstdlib>
+#include <string>
+#include <vector>
+
+#include "bench_utils.h"
+#include "lance/arrow/scanner.h"
+
+void Scan(const std::string& uri, const std::vector<std::string>& columns) {
+  auto dataset = OpenDataset(uri);
+  auto scan_builder = lance::arrow::ScannerBuilder(dataset);
+  assert(scan_builder.Project(columns).ok());
+  auto scanner = scan_builder.Finish().ValueOrDie();
+  fmt::print("Dataset count: {}\n", scanner->CountRows().ValueOrDie());
```

Comment:
does CountRows actually execute the scan?

Reply:
Yes, it is a virtual function that calls ScanBatch by default. Also this is observed when `SELECT count(1) FROM dataset` is not optimized in duckdb. 

We can change this code to use ScanBatch as well.